toplevel/topcommon.ml

```diff
@@ -380,3 +380,41 @@ let try_run_directive ppf dir_name pdir_arg =
             dir_name dir_type arg_type;
           false
   end
+
+(* Overriding exception printers with toplevel-specific ones *)
+
+let loading_hint_printer ppf err =
+  Symtable.report_error ppf err;
+  match err with
+  | Symtable.Undefined_global s ->
+    if !Sys.interactive then begin
+      let find_with_ext ext =
+        try Some (Load_path.find_uncap (s ^ ext)) with Not_found -> None
+      in
+      fprintf ppf
+        "@.Hint: @[\
+         This means a module's interface is loaded, \
+         but its implementation is not.@,";
+      begin match List.find_map find_with_ext [".cma"; ".cmo"] with
```

Comment:
Note: in general there is no reason for `foo.cma` to provide an implementation of the module `Foo`, and this makes this heuristic frustratingly incomplete. It is very wrong in the general case, but also correct surprisingly often (Dynlink, Unix, Graphics etc.). I would keep the code as you wrote it, but include a comment to point out that hinting at the .cma may be wrong.